lib/text/ui_text_displayer.js

```diff
@@ -189,6 +189,141 @@ shaka.text.UITextDisplayer = class {
     }
   }
 
+  /**
+   * Returns tokens describing start/end tags and string content from a cue.
+   *
+   * @param {string} str
+   * @return {!Array<{type: string, value: string}>}
+   * @private
+   */
+  tokenizeCueString_(str) {
+    if (!str) {
+      return [];
+    }
+
+    const tokens = [];
+
+    let currentValue = '';
+    for (let i = 0; i < str.length; i++) {
+      if (i+1 < str.length && str[i] === '<' && str[i+1] !== ' ') {
+        if (currentValue) {
+          // Flush the current value as we enter tag detection.
+          tokens.push({type: 'text', value: currentValue});
+        }
+        currentValue = '<';
+
+        // Seek to end of tag.
+        i++;
+        let wellTerminated = false;
+        for (;i < str.length; i++) {
+          currentValue += str[i];
+          if (str[i] === '>') {
+            wellTerminated = true;
+            break;
+          }
+        }
+
+        if (!wellTerminated || currentValue.length < 3) {
+          // No termination found, append literal string.
+          tokens.push({type: 'text', value: currentValue});
+        } else {
+          let tagType = 'startTag';
+          currentValue = currentValue.substr(1, currentValue.length - 2);
+          if (currentValue[0] == '/') {
+            currentValue = currentValue.substr(1); // Removes /
+            tagType = 'endTag';
+          }
+          if (currentValue) {
+            tokens.push({type: tagType, value: currentValue});
+          }
+        }
+        currentValue = '';
+      } else {
+        currentValue += str[i];
+      }
+    }
+
+    // Final flush.
+    if (currentValue) {
+      tokens.push({type: 'text', value: currentValue});
+    }
+
+    return tokens;
+  }
+
+  /**
+   * Returns a safe collection of html elements for display as a single cue.
+   *
+   * @param {!Array<{type: string, value: string}>} tokens
+   * @return {!Array<HTMLElement>}
+   * @private
+   */
+  buildCueElements_(tokens) {
+    const supportedTags = [
+      'i',
+      'b',
+      'u',
+    ];
+
+    const elements = [];
+
+    // Build a set of root elements from the token collection.
+    const stack = [];
+    let syntaxError = false;
+    for (const currentToken of tokens) {
+      if (currentToken.type === 'text') {
+        // Element is a text field.
+        const textElement = document.createElement('span');
+        textElement.textContent = currentToken.value;
+        if (stack.length) {
+          // Add to bottom stack element.
+          stack[stack.length - 1].appendChild(textElement);
+        } else {
+          // Add as standalone.
+          elements.push(textElement);
+        }
+      } else if (supportedTags.includes(currentToken.value)) {
+        // Element is a supported tag.
+        if (currentToken.type === 'startTag') {
```

Comment:
It would be safer to add an enum type so we don't have to worry about mistyping the enum values.  See [this](https://github.com/google/shaka-player/blob/fac29133c25b8c7955e8f2a2fa96c2b946731552/lib/net/networking_engine.js#L621) as an example.